src/compiler/checker.ts

```diff
@@ -6386,6 +6386,8 @@ namespace ts {
             // Only narrow when symbol is variable of type any or an object, union, or type parameter type
             if (node && symbol.flags & SymbolFlags.Variable) {
                 if (isTypeAny(type) || type.flags & (TypeFlags.ObjectType | TypeFlags.Union | TypeFlags.TypeParameter)) {
+                    const declaration = getDeclarationOfKind(symbol, SyntaxKind.VariableDeclaration);

```

Comment:
Yeah, `getDeclarationContainer` isn't a heavy walk. It's a static 2 or 4 references up (based on declaration kind) after walking out of the binding pattern in the declaration. I can't merge these walks as @mhegazy suggests because I'm not walking the same section of the tree twice (declaration -> up to declaration container vs use site -> up to declaration container - I mean, I could merge them, but I'd have to recur into all siblings at each level looking for the declaration, which seems _really bad_). If `k` is the stack of nodes between the file/module container and the node we're narrowing at (which is how many nodes I would walk in the simple case without looking for the declaration container of the symbol's declaration), and `n` is the number of nodes I walk up when computing narrowing before stopping at `top`, and `m` is the number of nodes I walk up when computing the declaration container to compute `top` (which is to say the recursive depth of the binding pattern), I'm wagering that walking over `m+n` is practically more performant than than `k`, given that `m` has a tendency to be very small and `k` has a tendency to be very large.

Now, if it turns out that the declaration container _is_ the top level of the namespace/module, then calculating it was extra work we didn't need to do as `n=k` and we've walked over `m+k` nodes rather than potentially just `k` nodes.

Really, this could be avoided if a declaration just kept a reference somewhere to its container, but that would be awkward, I suppose.

In effect, I'm trying to assert that `getDeclarationContainer` is close enough to constant time in all real-world cases, that its usage is warranted to be able to bail out of our narrowing walk up the tree early.